src/backend/chat/tools/fetch_url.py

```diff
@@ -0,0 +1,420 @@
+        content = message.get('content', '')
+        if content:
+            text_parts.append(content)
+    
+    return ' '.join(text_parts)
+
+
+def detect_url_in_conversation(conversation) -> list[str]:
+    """
+    Detect URLs present in the conversation messages.
+
+    Args:
+        conversation: The ChatConversation instance.
+
+    Returns:
+        list[str]: List of unique URLs found in the conversation.
+    """
+    if not conversation:
+        return []
+
+    found_urls = set()
+
+    def extract_urls_from_messages(messages):
+        if not messages:
+            return
+        for message in messages:
+            if not message:
+                continue
+            text_content = _extract_text_from_message(message)
+            if text_content and URL_PATTERN.search(text_content):
+                matches = URL_PATTERN.findall(text_content)
+                found_urls.update(matches)
+
+    # Get URLs from ui_messages and messages if present
+    if hasattr(conversation, 'ui_messages') and conversation.ui_messages:
+        extract_urls_from_messages(conversation.ui_messages)
+    if hasattr(conversation, 'messages') and conversation.messages:
+        extract_urls_from_messages(conversation.messages)
+
+    if found_urls:
+        logger.info("URL detected in messages: %s", found_urls)
+
+    return list(found_urls)
+
+
+async def _get_with_retry(
+    client: httpx.AsyncClient,
+    url: str,
+    max_attempts: int = 3,
+) -> httpx.Response:
+    """
+    Perform a GET request with randomized headers and a simple retry strategy.
+
+    We retry once on header-related HTTP status codes (e.g. 403, 429), each time
+    using a new random header set. Other errors are propagated immediately.
+    """
+    last_exception: httpx.HTTPStatusError | None = None
+
+    for attempt in range(max_attempts):
+        headers = _get_headers()
+        try:
+            response = await client.get(url, headers=headers)
+            response.raise_for_status()
+            return response
+        except httpx.HTTPStatusError as exc:
+            last_exception = exc
+            status_code = exc.response.status_code
+
+            # Only retry on codes that are likely related to headers / rate limits.
+            should_retry = status_code in (403, 429)
+            is_last_attempt = attempt >= max_attempts - 1
+
+            logger.debug(
+                "HTTP error %s for URL %s on attempt %s with headers %s (retry=%s)",
+                status_code,
+                url,
+                attempt + 1,
+                headers,
+                should_retry and not is_last_attempt,
+            )
+
+            if (not should_retry) or is_last_attempt:
+                raise
+
+    # Should not be reached, but keeps type-checkers happy.
+    if last_exception is not None:
+        raise last_exception
+
+    raise RuntimeError("Unexpected state in _get_with_retry")
+
+
+async def _store_in_rag_and_attachments(
```

Comment:
I think we should factorize the codebase, because this behavior is a copy/paste from the original workflow for attached documents.